webapp/src/ts/services/telemetry.service.ts

```diff
@@ -164,65 +116,117 @@ export class TelemetryService {
+      date: Number(parts[3]),
+    };
+  }
+
+  private async aggregate(db, dbName) {
+    const metadata = await this.generateMetadataSection(dbName);
+    const dbInfo = await this.dbService.get().info();
+    const reduceResult = await db.query(
+      { reduce: '_stats', map: (doc, emit) => emit(doc.key, doc.value) },
+      { group: true }
     );
 
-    return Promise
-      .all([
-        reduceQuery,
-        this.dbService.get().info(),
-        this.generateMetadataSection()
-      ])
-      .then(qAll => {
-        const reduceResult = qAll[0];
-        const infoResult = qAll[1];
-        const metadata = qAll[2];
-
-        const aggregateDoc: any = { type: 'telemetry' };
-
-        aggregateDoc.metrics = this.convertReduceToKeyValues(reduceResult);
-        aggregateDoc.metadata = metadata;
-        aggregateDoc._id = this.generateAggregateDocId(aggregateDoc.metadata);
-        aggregateDoc.device = this.generateDeviceStats();
-        aggregateDoc.dbInfo = infoResult;
-
-        return this.dbService
-          .get({ meta: true })
-          .put(aggregateDoc)
-          .catch(err => {
-            if (err.status === 409) {
-              return this.storeConflictedAggregate(aggregateDoc);
-            }
-
-            throw err;
-          });
-      });
+    const aggregateDoc = {
+      _id: this.generateAggregateDocId(metadata),
+      type: this.TELEMETRY_PREFIX,
+      metrics: this.convertReduceToKeyValues(reduceResult),
+      device: this.generateDeviceStats(),
+      metadata,
+      dbInfo,
+    };
+
+    try {
+      await this.dbService
+        .get({ meta: true })
+        .put(aggregateDoc);
+    } catch (error) {
+      if (error.status === 409) {
+        return this.storeConflictedAggregate(aggregateDoc);
+      }
+      throw error;
+    }
+  }
+
+  /**
+   * Moment when the aggregation starts (i.e. the beginning of the current day)
+   */
+  private getToday(): TodayMoment {
+    const today = moment().startOf('day');
+    return {
+      today,
+      formatted: today.format(this.DATE_FORMAT),
+    };
+  }
+
+  private async getCurrentTelemetryDB(today: TodayMoment) {
+    const telemetryDBs = await this.getTelemetryDBs();
+    let currentDB = telemetryDBs?.find(db => db.includes(today.formatted));
+
+    if (!currentDB) {
+      currentDB = this.generateTelemetryDBName(today);
+    }
+
+    return this.windowRef.PouchDB(currentDB); // Avoid angular-pouch as digest isn't necessary here
+  }
+
+  private storeIt(db, key, value) {
+    return db.post({
+      key: key,
+      value: value,
+      date_recorded: Date.now(),
+    });
   }
 
-  private reset(db) {
-    window.localStorage.removeItem(this.DB_ID_KEY);
-    window.localStorage.removeItem(this.FIRST_AGGREGATED_DATE_KEY);
+  private async submitIfNeeded(today: TodayMoment, telemetryDBs: string[] = []) {
+    for (const dbName of telemetryDBs) {
+      if (this.isAggregationRunning || dbName.includes(today.formatted)) {
```

Comment:
> I wonder if it is here as a locking mechanism to try and avoid multiple calls simultaneous calls to the telemetry record function from doing duplicate aggregation work.

Correct, when calling `telemetry.record(...)` there's a risk of getting this exception: `Failed to execute 'transaction' on 'IDBDatabase': The database connection is closing.`. That's what we're trying to avoid with `isAggregationRunning`, because the other option is to always `await` for `telemetry.record(...)`. So, it's safer to do it here.

> would it make more sense to move the isAggregationRunning logic outside of the for-of loop?

Yes, that makes a lot of sense! Thanks

Reply:
I've added a comment to clarify this